### PR TITLE
io: add `copy_buf`

### LINKED ISSUE #2797: Add copy_buf
Add `copy_buf` function like [the one from futures](https://docs.rs/futures/0.3/futures/io/fn.copy_buf.html), but with our async io traits.

### PR DESCRIPTION
**Pull request by zaharidichev**
## Motivation
Add copy_buf function like the one from futures, but with tokio's async io traits.

## Solution
Add `io::copy_buf` function.
Fix #2797

Signed-off-by: Zahari Dichev <zaharidichev@gmail.com>

### CODE REVIEW

File: tokio/src/io/util/copy_buf.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /// writer.
+    ///
+    /// This struct is generally created by calling [`copy_buf`][copy_buf]. Please
+    /// see the documentation of `copy_buf()` for more details.
+    ///
+    /// [copy_buf]: copy_buf()
+    #[derive(Debug)]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct CopyBuf<'a, R: ?Sized, W: ?Sized> {
+        reader: &'a mut R,
```
Comment (taiki-e) [inline tokio/src/io/util/copy_buf.rs:17]: FYI: `futures`'s `copy_buf` [takes the reader by value](https://github.com/rust-lang/futures-rs/blob/8f20432c0271a86e7cc31fb0612e690157e8c74d/futures-util/src/io/copy_buf.rs#L51) to allow `!Unpin` reader. (This is only done in the reader as only the reader is exhausted by this operation.)

Comment (Darksonn) [reply tokio/src/io/util/copy_buf.rs:17]: On the other hand, taking it by reference is consistent with Tokio's `copy`. In either case, `!Unpin` readers can still be used here if combined with `tokio::pin!`.

File: tokio/src/io/util/copy_buf.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    /// ```
+    /// use tokio::io;
+    ///
+    /// # async fn dox() -> std::io::Result<()> {
+    /// let mut reader: &[u8] = b"hello";
+    /// let mut writer: Vec<u8> = vec![];
+    ///
+    /// io::copy_buf(&mut reader, &mut writer).await?;
+    ///
+    /// assert_eq!(&b"hello"[..], &writer[..]);
```
Comment (Darksonn) [inline tokio/src/io/util/copy_buf.rs:49]:
```suggestion
    /// assert_eq!(b"hello", &writer[..]);
```

**Comment by carllerche** [comment]
This is fine w/ me, but based on #2428, it looks like `AsyncBufRead` is going to be moved to `tokio-util`. Maybe we should wait for that to happen before merging this PR? Thoughts @seanmonstar

File: tokio/src/io/util/copy_buf.rs (hunk 1 above)
Comment (taiki-e) [reply tokio/src/io/util/copy_buf.rs:17]: > On the other hand, taking it by reference is consistent with Tokio's `copy`.

Yeah, if `copy_buf` takes the reader by value, `copy` also needs to take the reader by value for consistency.

**Comment by Darksonn** [comment]
Note that the `AsyncBufRead` thing was postponed for now.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**